File: c/grains/grains.c
#include "grains.h"

#define MAX_SQUARES 64

uint64_t square(uint8_t index) {
  if (index > MAX_SQUARES) {
    return 0;
  } else if (index <= 2) {
    return index;
  }

  uint64_t sq = 2;
  for (int i = 2; i < index; i++) {
    sq *= 2;
  }
  return sq;
}

uint64_t total(void) {
  uint64_t sq = 2;
  uint64_t sum = 3;  // 1st + 2nd
  for (int i = 1; i <= MAX_SQUARES; i++) {
    sq *= 2;
    sum += sq;
  }
  return sum;
}

File: csharp/grains/Grains.cs
public static class Grains
{
  private const int SQUARES_NUMBER = 64;

  public static ulong Square(int n)
  {
    ArgumentOutOfRangeException.ThrowIfLessThanOrEqual(n, 0);
    ArgumentOutOfRangeException.ThrowIfGreaterThan(n, SQUARES_NUMBER);

    return (ulong)Math.Pow(2, n - 1);
  }

  public static ulong Total()
  {
    checked
    {
      try
      {
        return 2UL * Square(SQUARES_NUMBER);
      }
      catch (OverflowException)
      {
        return ulong.MaxValue;
      }
    }
  }
}
